cxx4opencl/diff2cxx.txt

```diff
@@ -12,8 +12,22 @@ however, there are some differences that are documented in this section.
 
 The following {cpp} language features are not supported:
 
-  * Virtual functions
-  * Exceptions
-  * `dynamic_cast` operator
-  * Non-placement `new`/`delete` operators
-  * Standard C++ libraries
+  * Virtual functions ({cpp}17 `[class.virtual]`);
+  * References to functions including member functions ({cpp}17 `[class.mfct]`);
+  * Pointers to class member functions (in addition to the regular non-member
+    functions that are already restricted in OpenCL C);
+  * Exceptions ({cpp}17 `[except]`);
+  * `dynamic_cast` operator ({cpp}17 `[expr.dynamic.cast]`);
+  * Non-placement `new`/`delete` operators ({cpp}17
+    `[expr.new]`/`[expr.delete]`);
+  * Standard C++ libraries ({cpp}17 `[library]`).
+
+Thread-safe initialization of static local objects is not guaranteed ({cpp}17
+`[stmt.dcl]`) and whether an implementation provides thread-safety is
+indicated by the presence of the `__cpp_threadsafe_static_init` feature
+test macro.
```

Comment:
I think this is the first time that a C++ feature macro has been referenced in C++ for OpenCL, though I don't know if more from C++ spec are inherited automatically.

For OpenCL C feature test macros we have the `CL_DEVICE_OPENCL_C_FEATURES` device query.  Should a new revision of the `cl_ext_cxx_for_opencl` extension add a `CL_DEVICE_CXX_FOR_OPENCL_FEATURES` query?

It is unlikely that this can be achieved in this PR, so if there is an action to take it could just be an issue for now.

Is threadsafe the correct name here too?  Work-items aren't normally referred to as threads.  I do realize that this name comes from C++ though, so it would be easy to argue that keeping it the same is consistent.

Reply:
> I think this is the first time that a C++ feature macro has been referenced in C++ for OpenCL, though I don't know if more from C++ spec are inherited automatically.

Yeah, it is assumed that everything that is not covered in this section is automatically inherited from C++. However, the feature test macros are not in C++17 though. So I have added a footnote to make clear that this is a C++20 functionality.

> For OpenCL C feature test macros we have the CL_DEVICE_OPENCL_C_FEATURES device query. Should a new revision of the cl_ext_cxx_for_opencl extension add a CL_DEVICE_CXX_FOR_OPENCL_FEATURES query?
> 
> It is unlikely that this can be achieved in this PR, so if there is an action to take it could just be an issue for now.
> 

We haven't ported C++ for OpenCL to OpenCL 3.0 yet, so there were no feature test macros provided so far. But I think it would indeed be good to record this for the future.

> Is threadsafe the correct name here too? Work-items aren't normally referred to as threads. I do realize that this name comes from C++ though, so it would be easy to argue that keeping it the same is consistent.

Good point. We don't have any equivalent term for work-items so I have rephrased it now to:

```
Simultaneous initialization of static local objects performed by
different work-items is not guaranteed to be free from race-condition.
Whether an implementation provides such a guarantee is indicated by the
presence of the `__cpp_threadsafe_static_init` feature test
macro{fn-feature-macro}.
```

Thanks for your feedback!